File: ex03/java/kkw5240.java
package ex03.java;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Ex03 {
    private static Scanner scanner;
    public static void main(String[] args) {
        scanner = new Scanner(System.in);

        Quotes quotes = new Quotes();

        while (isEmptyInput()) {
            Quote quote = makeQuote();
            quote.printQuote();

            quotes.add(quote);
        }
        quotes.printQuotes();

        scanner.close();
    }

    private static Quote makeQuote() {
        String message = getMessage();
        String person = getPerson();

        return new Quote(person, message);
    }

    private static boolean isEmptyInput() {
        return scanner.nextLine().isEmpty();
    }

    private static String getMessage() {
        System.out.print("What is the quote? ");
        return scanner.nextLine();
    }

    private static String getPerson() {
        System.out.print("Who said it? ");
        return scanner.nextLine();
    }

    private static class Quote {
        private final String person;
        private final String message;

        public Quote(String person, String message) {
            this.person = person;
            this.message = message;
        }

        public void printQuote() {
            System.out.println(this.person + " says, \"" + this.message + ".\"");
        }
    }

    private static class Quotes {
        private final List<Quote> quotes;

        public Quotes() {
            this.quotes = new ArrayList<>();
        }

        public void add(Quote quote) {
            this.quotes.add(quote);
        }

        public void printQuotes() {
            this.quotes.forEach(Quote::printQuote);
            System.out.println();
        }
    }
}

File: ex03/go/kkw5240.go
package main

import "fmt"

/**
 * 작성자 : 김규완
 * 실행방법 : go kkw5240.go
 * 주의사항 : 기본풀이까지만 했습니다.
 */
func main() {
	var quotes []*Quote

	for {
		quote := &Quote{
			quote:  promptQuote(),
			person: promptPerson(),
		}
		if !quote.Validate() {
			return
		}

		quote.PrintQuote()

		fmt.Println("\n==============================================")

		quotes = append(quotes, quote)
		for _, q := range quotes {
			q.PrintQuote()
		}
	}
}

func promptQuote() string {
	fmt.Print("What is the quote? ")

	var quote string
	_, err := fmt.Scanln(&quote)
	if err != nil {
		return ""
	}

	return quote
}

func promptPerson() string {
	fmt.Print("Who said it? ")

	var person string
	_, err := fmt.Scanln(&person)
	if err != nil {
		return ""
	}

	return person
}

type Quote struct {
	person string
	quote  string
}

func (q *Quote) PrintQuote() {
	fmt.Println(q.buildMessage())
}

func (q *Quote) buildMessage() string {
	return fmt.Sprintf("%s says, \"%s.\"", q.person, q.quote)
}

func (q *Quote) Validate() bool {
	return len(q.quote) > 0 && len(q.person) > 0
}
